documentation/ArrayExpressionLoweringToFIR.md

```diff
@@ -0,0 +1,315 @@
+- compute `x + tmp1 + z`
+With this alternative, transformational such as `transpose` are still treated as `foo`.
+Pros: Generates less FIR, the original expression structure is better preserved inside the loop.
+Cons: Slightly more complex lowering, we need to keep track of the loop where we want to emit the "inlined" operations and where to insert
+the computation of sub-expressions that are not "inlined".
+
+#### One "Affine" Elemental Expression Sub-Tree - One Loop
+Same as before, but also apply transformationals as index re-mapping instead of computing it outside of the loop.
+For instance `x + transpose(y) + z` becomes a loop computing each `x(i, j) + y(j, i) + z(i, j)`.
+Pros: Generates minimal FIR. Not much more complex that the previous one (mostly the same thing).
+Cons: Requires more knowledge about transformationals, an index mapping state must be propagated while lowering.
+
+
+## Examples
+Below are illustrations of Fortran to FIR lowering based on the above design (with one node, one loop). 
+
+### Simple examples
+Fortran:
+```
+  real :: x(4), y(4)
+  y = x
+```
+Here, `x` expression should not be split into a loop.
+
+FIR:
+```
+ %1 = fir.load %x // that is the whole expression evaluation
+ // This is actually the assignment lowering
+ %fir.store %1 to %y
+```
+Question: This is more related to assignment lowering, but what will decide whether `%1`
+is has to be copied inside a temporary or not (e.g. in case of pointers that may overlap or not).
+Is it implicit in here because `%1` requires a storage later and
+FIR will determine whether `%y` can directly be used ? 
+### Outline scalar sub-expression outside of the loop
+Fortran:
+```
+  real :: x(4), y(4)
+  y = x + cos(1.)
+```
+FIR:
+```
+  %tmp = fir.alloca !fir.array<4:f32>
+  %cst = constant 1.000000e+00 : f32
+  %1 = fir.call "cos" %cst
+  fir.loop %i = 1 to 4 {
+    %2 = fir.extract_value %x, %i
+    %3 = fir.addf %2, %1
+    %4 = fir.coordinate_of %tmp, %i
+    fir.store %3 to %4 
+  }
+  %5 = fir.load %tmp // This is the expression result
+  // The assignment part is not part of the expression lowering
+  fir.store %5 to %y
+```
+### Dynamic shape example
+A more complex case where the shape is not a compile time constant.
+Fortran:
+```
+   real, allocatable :: x(:), y(:)
+   y = x + cos(1.)
+```
+FIR:
+```
+  %_,_, %1 = fir.box_dims %x
+  %cst_0 = constant 0 : i64
+  %3 = cmpi eq %1 %cst_0
+  cond_br %3 ^bnullshape ^bb1
+  ^bb1:
+    %tmp = fir.allocmem !fir.array<?:f32> %1
+    %cst = constant 1.000000e+00 : f32
+    %cosres = fir.call "cos" %cst
+    %lbx1, _, _ = fir.box_dims %x
+    %cst_1 = constant 1 : i64
+    %lbx0 = subi lbx1 %cst_1
+    fir.loop %i = 1 to %1 {
+
+      %ix = subi %i %lbx0
+      %2 = fir.coordinate_of %x, %i
+      %3 = fir.load %2
+      %4 = fir.addf %3, %cosres
+      %5 = fir.coordinate_of %tmp, %i
+      fir.store %4 to %5
+    }
+    %6 = fir.load %tmp
+    fir.freemem %tmp // are we allowed to do this so soon?
```

Comment:
Probably not a good idea to get rid of the backing store while the array value is live.

Reply:
Agreed, this will require some work to propagate it until the last expression consumes related its ssa value and ensure it is then freed.

Reply:
Hopefully scopes will help us here.